My Site Dashboard (Phase 2): Refine Quick Start card design

momo-ozawa opened the pull request:
Part of #17877 

## Description
- Refines Quick Start dashboard card UI
- Handles a11y for the Quick Start dashboard card

## How to test

### Quick Start dashboard card UI

Make sure the MSD feature flag is enabled

1. First we need to reset Quick Start -- Remove "Next steps" from your site menu view if you have it showing. Otherwise, continue to the next step
2. Enable Quick Start from App Settings > Debug 
3. Tap "Home" on the segmented control to display the dashboard
4. Go through all tours on both the "Customize Your Site" and the "Grow Your Audience" checklist
5. ✅ Make sure that the circular progress view updated when a tour is completed
6. ✅ Make sure that the subtitle label is updated when a tour is completed
7. ✅ Make sure that the title label is crossed out after completing all tours for a checklist

![Simulator Screen Shot - iPhone 13 - 2022-02-28 at 17 33 20](https://user-images.githubusercontent.com/6711616/156030891-9e72bf68-6920-4622-93a8-e61690d1a1f7.png) | ![Simulator Screen Shot - iPhone 13 - 2022-02-28 at 17 34 24](https://user-images.githubusercontent.com/6711616/156030895-f15cf580-62bc-4e59-b6aa-da25cebe66b6.png) | ![Simulator Screen Shot - iPhone 13 - 2022-02-28 at 17 35 06](https://user-images.githubusercontent.com/6711616/156030899-843673f5-d3e0-4071-8609-7de72938be3c.png)
-- | -- | --

### A11y

Make sure the MSD feature flag is enabled

1. Turn on VoiceOver
2. Go to My Site > Home tab
3. Tap on "Customize Your Site"
4. ✅ The title, subtitle, and hint is read out loud
5. 3. Tap on "Grow Your Audience"
4. ✅ The title, subtitle, and hint is read out loud


https://user-images.githubusercontent.com/6711616/156031035-681245e3-b841-4190-adac-2105d4252e4f.mov


## Regression Notes
1. Potential unintended areas of impact
n/a

6. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

7. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: QuickStartChecklistView

Commit:
Add: update checklist view when tours are completed

Commit:
Add: progress indicator view to checklist view

Commit:
Add: a11y for quick start checklist view

hassaanelgarem commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Start/DashboardQuickStartCardCell.swift:
```diff
@@ -13,22 +10,28 @@ final class DashboardQuickStartCardCell: UICollectionViewCell, Reusable, BlogDas
+    private lazy var growChecklistView: QuickStartChecklistView = {
+        let view = QuickStartChecklistView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.accessibilityHint = Strings.customizeHint
```
This should be `growHint`. But do we actually need that? 
Looks like this is handled in `QuickStartChecklistView.configure`

momo-ozawa replied:
Fixed! [a0b4666](https://github.com/wordpress-mobile/WordPress-iOS/pull/18044/commits/a0b4666b046acb8659b89b73752193ebe38f9228)

hassaanelgarem commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Start/QuickStartChecklistView.swift:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    deinit {
+        startObservingQuickStart()
```
I think you meant here to call `stopObservingQuickStart` instead?

momo-ozawa replied:
OOPS. 🙈 

Fixed! 9f7f7c2

hassaanelgarem commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Start/QuickStartChecklistView.swift:
```diff
@@ -0,0 +1,159 @@
+    private func setupViews() {
+        addSubview(mainStackView)
+
+        NSLayoutConstraint.activate([
```
Just a suggestion that you can ignore:

What about using `pinSubviewToAllEdges(_ subview: UIView, insets: UIEdgeInsets)` here?

momo-ozawa replied:
Fixed! [cf033d4](https://github.com/wordpress-mobile/WordPress-iOS/pull/18044/commits/cf033d4a49e7c9285fa207fcf91ebc2df25fb416)

hassaanelgarem reviewed (commented):
I’ve faced an issue while testing this PR, but I’m not sure if it’s specific to this PR or to my environment. When running the app, once I go to the Home tab, the app just freezes and the debugger keeps logging the below message indefinitely: 
`WordPress[4835:2173661] [UICollectionViewRecursion] cv == 0x11883c200 Disabling recursion trigger logging`
Looks like an infinite loop or something. This doesn't happen consistently. And going back to trunk fixes it.

The UI works as expected, I've just noticed a few minor issues/proposed enhancements:
- I noticed that for QuickStart in Site Menu when a checklist is complete, not only is the title crossed out but also the font color is updated to reflect it’s complete. Not sure if we should follow that for the dashboard’s QuickStart as well or not.
- When switching from Dark to Light mode or vice versa, the progress’s track color is not updated.
<img src="https://user-images.githubusercontent.com/25306722/156395830-d17177a9-9c84-48c6-aed1-877ccdb7c0dc.PNG" width=49%> <img src="https://user-images.githubusercontent.com/25306722/156395910-688665ea-8582-4d83-9fbf-bbfc397ac9ad.PNG" width=49%>
- When going through the Customize Your Site tour, finishing step 4 is not reflected in QuickStart Dashboard card. It shows “3 of 6 completed” while the My Site QuickStart shows the correct progress: “4 of 6 completed”

https://user-images.githubusercontent.com/25306722/156396410-f7f649dd-32f3-4040-b040-135df8081909.MP4

As for the a11y:
- The title and subtitle are read out loud as expected ✅ 
- However, the hint is not read. It seems to be added to an inaccessible element.

Finally, I've left some minor comments on the code 👍

Commit:
Fix: stop observing quick start in deinit

Commit:
Fix: no need to set a11y hint directly

The a11y hint is set via configure

leandroalonso commented:
@hassaanelgarem 

> When running the app, once I go to the Home tab, the app just freezes and the debugger keeps logging the below message indefinitely: 
WordPress[4835:2173661] [UICollectionViewRecursion] cv == 0x11883c200 Disabling recursion trigger logging
Looks like an infinite loop or something. This doesn't happen consistently. And going back to trunk fixes it.

I think this problem is related to [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/17921#pullrequestreview-877877396) (we also have this as a task on #17873.

What I believe is happening is that the posts card is triggering a layout update at the same time as the QuickStart cell is also doing, this is causing an infinite loop. We'll probably have to find another way to resize the Posts card cell.

To test if this is really the issue you can just comment the line inside `BlogDashboardViewController.updateCollectionViewHeight()`

momo-ozawa commented:
> When going through the Customize Your Site tour, finishing step 4 is not reflected in QuickStart Dashboard card. It shows “3 of 6 completed” while the My Site QuickStart shows the correct progress: “4 of 6 completed”

@hassaanelgarem I think I know why this is happening. If you got to Pages, do you have a home page like this?
<img src="https://user-images.githubusercontent.com/6711616/156409328-edbf914e-4c97-44a4-a81e-8a5e609a6726.jpeg" width=200>

I think the issue you pointed above happens when you **don't** have a home page

Commit:
Refactor: use edge insets to pin subview

hassaanelgarem commented:
> To test if this is really the issue you can just comment the line inside

@leandroalonso That fixed it indeed 👍 

> I think the issue you pointed above happens when you don't have a home page

@momo-ozawa You are correct here. Retested after setting a homepage and works as expected 👍 
I know this is outside this PR's scope, but shouldn't we avoid showing this step if the user doesn't have a homepage? 🤔

Commit:
Fix: a11y for checklist view

momo-ozawa commented:
> I noticed that for QuickStart in Site Menu when a checklist is complete, not only is the title crossed out but also the font color is updated to reflect it’s complete. Not sure if we should follow that for the dashboard’s QuickStart as well or not.

I tried to match the Android design -- the Quick Start card font color doesn't change when the tasks are completed @osullivanchris, what do you think? Should we update the font color when the title is crossed out?

<img width="369" alt="Screen Shot 2022-03-02 at 18 06 08" src="https://user-images.githubusercontent.com/6711616/156421697-1fc6f153-b6c6-401a-a83e-db5631317108.png">

> When switching from Dark to Light mode or vice versa, the progress’s track color is not updated.

~Hmm! I'm looking at your screenshots and they look different from mine. I'm using `.listBackground` for the track color, which points to `.systemGroupedBackground`~

Fixed! 44a5b83

You’re 100% right, the color doesn’t change. I thought the grays looked slightly different between light / dark mode, but then I realized that since it’s pointing to `.systemGroupedBackground` it should look black on dark mode.

The reason the color doesn't update is because [CGColor doesn't have UIColor’s dynamic behavior](https://www.jessesquires.com/blog/2020/03/23/implementing-dark-mode-with-cgcolor/). Great catch! I also updated the progress track color to use `.separator` instead, because I thought `.systemGroupedBackground` looked too dark in dark mode.

Light | Dark
-- | --
<img src="https://user-images.githubusercontent.com/6711616/156438642-6acc074d-6c4f-42c6-9cab-7e6386e05be1.PNG" width=200> | <img src="https://user-images.githubusercontent.com/6711616/156438636-6244ba16-781b-4a37-8071-d5f237f14c4c.PNG" width=200>



> As for the a11y:
> 
> - The title and subtitle are read out loud as expected ✅
> - However, the hint is not read. It seems to be added to an inaccessible element.

Fixed! 2afcb78

<details>
 <summary>Video below (sound on to hear VoiceOver)</summary>
 
 https://user-images.githubusercontent.com/6711616/156417590-efef8439-0bf9-4625-ba95-2498156eed82.mov
 
</details>

>  I know this is outside this PR's scope, but shouldn't we avoid showing this step if the user doesn't have a homepage? 🤔

Yeah, I agree. I created an issue for this: https://github.com/wordpress-mobile/WordPress-iOS/issues/18066

Commit:
Fix: update layer cgcolors when switching between light / dark mode

https://www.jessesquires.com/blog/2020/03/23/implementing-dark-mode-with-cgcolor/

osullivanchris commented:
> I tried to match the Android design -- the Quick Start card font color doesn't change when the tasks are completed @osullivanchris, what do you think? Should we update the font color when the title is crossed out?

@momo-ozawa sure, I think the secondary label color de-emphasises it nicely. But given its already crossed-out and has a completed progress circle, its no big deal.

hassaanelgarem commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Quick Start/QuickStartChecklistView.swift:
```diff
@@ -0,0 +1,159 @@
+        let tap = UITapGestureRecognizer(target: self, action: #selector(didTap))
+        addGestureRecognizer(tap)
+
+        isAccessibilityElement = true
+        accessibilityTraits = .button
```
A tiny suggestion would be to group all accessibility logic in one helper and call it from `configure`. Currently, it's split over three places. Feel free to ignore this though 👍

momo-ozawa replied:
Fixed! a2e4c5f

I still need to set the accessibilityLabel in `updateViews()` -- otherwise, the "x out of y completed" doesn't get updated.

hassaanelgarem reviewed (approved):
Retested and all looks good to me 🚀

Commit:
Refactor: configure accessibility in one place

Commit:
Update: change title label font color when all tours are complete
